### PR TITLE
Sort by length

### PR DESCRIPTION
**Pull request by divyadaglia-unbxd**

### CODE REVIEW

File: unbxdAutosuggest.js [hunk 1]
```diff
@@ -1488,6 +1573,11 @@ var unbxdAutoSuggestFunction = function ($, Handlebars, undefined) {
 				key = 'prepare' + key + 'HTML';
 				html = html + self[key]();
 			});
+
+			if (this.options.sortByLength) {
+				html = html + self['prepareSortedSuggestionsHTML']();
+			}
+
```
Comment (anantheshadiga) [inline unbxdAutosuggest.js]: Please add a test case, Else everything looks good

Comment (divyadaglia-unbxd) [reply unbxdAutosuggest.js]: done

File: unbxdAutosuggest.js [hunk 2]
```diff
@@ -1372,24 +1440,32 @@ var unbxdAutoSuggestFunction = function ($, Handlebars, undefined) {
...
 		, preparetopQueriesHTML: function () {
-			return '{{#if data.TOP_SEARCH_QUERIES}}'
-				+ (this.options.topQueries.header ? '<li class="unbxd-as-header">' + this.options.topQueries.header + '</li>' : '')
-				+ '{{#each data.TOP_SEARCH_QUERIES}}'
-				+ '<li class="unbxd-as-keysuggestion" data-type="{{type}}" data-index="{{@index}}" data-value="{{autosuggest}}">'
-				+ (this.options.topQueries.tpl ? this.options.topQueries.tpl : this.default_options.topQueries.tpl)
-				+ '</li>'
-				+ '{{/each}}'
-				+ '{{/if}}';
+			if (this.options.sortByLength) {
```
[the remark this answers is not in the document]
Comment (divyadaglia-unbxd) [reply unbxdAutosuggest.js]: done

File: unbxdAutosuggest.js [hunk 3]
```diff
@@ -1372,24 +1440,32 @@ var unbxdAutoSuggestFunction = function ($, Handlebars, undefined) {
...
 		, preparekeywordSuggestionsHTML: function () {
-			return '{{#if data.KEYWORD_SUGGESTION}}'
-				+ (this.options.keywordSuggestions.header ? '<li class="unbxd-as-header">' + this.options.keywordSuggestions.header + '</li>' : '')
-				+ '{{#each data.KEYWORD_SUGGESTION}}'
-				+ '<li class="unbxd-as-keysuggestion" data-value="{{autosuggest}}" data-index="{{@index}}" data-type="{{type}}"  data-source="{{source}}">'
-				+ (this.options.keywordSuggestions.tpl ? this.options.keywordSuggestions.tpl : this.default_options.keywordSuggestions.tpl)
-				+ '</li>'
-				+ '{{/each}}'
-				+ '{{/if}}';
+			if (this.options.sortByLength) {
```
[the remark this answers is not in the document]
Comment (divyadaglia-unbxd) [reply unbxdAutosuggest.js]: done

File: unbxdAutosuggest.js [hunk 4]
```diff
@@ -1411,6 +1487,15 @@ var unbxdAutoSuggestFunction = function ($, Handlebars, undefined) {
 				+ '{{/data.POPULAR_PRODUCTS}}'
 				+ '{{/if}}';
 		}
+		, prepareSortedSuggestionsHTML: function () {
+			return '{{#if data.SORTED_SUGGESTIONS}}'
+				+ '{{#each data.SORTED_SUGGESTIONS}}'
+				+ '<li class="unbxd-as-sorted-suggestion" data-type="{{type}}" data-index="{{@index}}" data-value="{{autosuggest}}">'
```
[the remark this answers is not in the document]
Comment (divyadaglia-unbxd) [reply unbxdAutosuggest.js]: changed the implementation to take care of index

File: unbxdAutosuggest.js [hunk 5]
```diff
@@ -1486,6 +1566,9 @@ var unbxdAutoSuggestFunction = function ($, Handlebars, undefined) {
 				//main zero side not zero
 				if ((mainlen == 0) && (sidelen != 0)) {
 					this.options.sideTpl.forEach(function (key) {
+						if (self.options.sortByLength && (key == 'topQueries' || key == 'keywordSuggestions')) {
```
Comment (ganeshparsads) [inline unbxdAutosuggest.js]: Actually, we can use `key !== "SORTED_SUGGESTIONS"`. Overall bundling code size will reduce.

File: unbxdAutosuggest.js [hunk 6]
```diff
@@ -1494,6 +1577,9 @@ var unbxdAutoSuggestFunction = function ($, Handlebars, undefined) {
 					if (sidelen != 0) {
 						html = '<ul class="unbxd-as-sidecontent">';
 						this.options.sideTpl.forEach(function (key) {
+							if (self.options.sortByLength && (key == 'topQueries' || key == 'keywordSuggestions')) {
```
Comment (ganeshparsads) [inline unbxdAutosuggest.js]: actually, we can use `key !== "SORTED_SUGGESTIONS"`. Overall bundling code size will reduce.

File: unbxdAutosuggest.js [hunk 7]
```diff
@@ -1503,9 +1589,17 @@ var unbxdAutoSuggestFunction = function ($, Handlebars, undefined) {
 
 			}
 			this.options.mainTpl.forEach(function (key) {
+				if (self.options.sortByLength && (key == 'topQueries' || key == 'keywordSuggestions')) {
```
Comment (ganeshparsads) [inline unbxdAutosuggest.js]: Here also we can use `key !== "SORTED_SUGGESTIONS"`.

File: unbxdAutosuggest.js [hunk 8]
```diff
@@ -1281,6 +1329,25 @@ var unbxdAutoSuggestFunction = function ($, Handlebars, undefined) {
...
+			this.currentResults.SORTED_SUGGESTIONS.sort(function (a, b) {
+				if (a.autosuggest.length - b.autosuggest.length == 0) {
+					return a.autosuggest.localeCompare(b.autosuggest);
+				}
+				else {
+					return a.autosuggest.length - b.autosuggest.length;
+				}
+			});
+			this.currentResults.IN_FIELD.sort(function (a, b) {
+				if (a.autosuggest.length - b.autosuggest.length == 0) {
```
Comment (ganeshparsads) [inline unbxdAutosuggest.js]: Repeated `sort` function.

File: unbxdAutosuggest.js [hunk 9]
```diff
@@ -1281,6 +1329,25 @@ var unbxdAutoSuggestFunction = function ($, Handlebars, undefined) {
 			}
 
 		}
+		, sortSuggestionsBylength: function () {
+			this.currentResults.SORTED_SUGGESTIONS = this.currentResults.KEYWORD_SUGGESTION.concat(this.currentResults.TOP_SEARCH_QUERIES);
+			this.currentResults.SORTED_SUGGESTIONS.sort(function (a, b) {
```
Comment (ganeshparsads) [inline unbxdAutosuggest.js]: Hoist common `sort` function and use it. Same code has been used twice.

File: unbxdAutosuggest.js [hunk 10]
```diff
@@ -1281,6 +1329,25 @@ var unbxdAutoSuggestFunction = function ($, Handlebars, undefined) {
...
+				if (a.autosuggest.length - b.autosuggest.length == 0) {
```
Comment (ganeshparsads) [inline unbxdAutosuggest.js]: use `===` for being on the safer side.

Below two are different:
`"0" == 0` => returns true;
`"0" === 0` => returns false;

`===` returns true if both operands are of the same type and contain the same value.

File: unbxdAutosuggest.js [hunk 11]
```diff
@@ -1397,7 +1468,7 @@ var unbxdAutoSuggestFunction = function ($, Handlebars, undefined) {
 				+ (this.options.keywordSuggestions.tpl ? this.options.keywordSuggestions.tpl : this.default_options.keywordSuggestions.tpl)
 				+ '</li>'
 				+ '{{/each}}'
-				+ '{{/if}}';
+				+ '{{/if}}';
```
Comment (ganeshparsads) [inline unbxdAutosuggest.js]: Remove extra spaces.

File: unbxdAutosuggest.js [hunk 12]
```diff
@@ -1407,7 +1478,7 @@ var unbxdAutoSuggestFunction = function ($, Handlebars, undefined) {
 				+ (this.options.topQueries.tpl ? this.options.topQueries.tpl : this.default_options.topQueries.tpl)
 				+ '</li>'
 				+ '{{/each}}'
-				+ '{{/if}}';
+				+ '{{/if}}';
```
Comment (ganeshparsads) [inline unbxdAutosuggest.js]: Remove extra spaces.

File: unbxdAutosuggest.js [hunk 13]
```diff
@@ -1246,6 +1254,29 @@ var unbxdAutoSuggestFunction = function ($, Handlebars, undefined) {
...
+					for (var j = 0; j < ins[i].length; j++) {
+						sortedInfields[k] = {
+							filterName: i,
+							filterValue: ins[i][j]
+						}
+						k++;
+					}
+				}
+
+				sortedInfields.sort(function (a, b) {
```
Comment (ganeshparsads) [inline unbxdAutosuggest.js]: Repeated `sort` function.

File: unbxdAutosuggest.js (hunk 5 above)
Comment (anantheshadiga) [reply unbxdAutosuggest.js]: How does logically you explain `key !== "SORTED_SUGGESTIONS"` ? What if tomorrow, we have more suggestion, let's consider the GroupOfSuggestion? I mean, I can understand this `(self.options.sortByLength && (key == 'topQueries' || key == 'keywordSuggestions'))`

Comment (anantheshadiga) [reply unbxdAutosuggest.js]: I mean, I dont see any problem as such that

File: unbxdAutosuggest.js (hunk 10 above)
Comment (anantheshadiga) [reply unbxdAutosuggest.js]: How does `a.autosuggest.length - b.autosuggest.length` return `"0"` ?

**Comment by ganeshparsads** [comment]
@divyadaglia-unbxd could you please correct this PR.

File: unbxdAutosuggest.js (hunk 5 above)
Comment (divyadaglia-unbxd) [reply unbxdAutosuggest.js]: @ganeshparsads Using !sorted suggestions will include infields too

File: unbxdAutosuggest.js (hunk 7 above)
Comment (divyadaglia-unbxd) [reply unbxdAutosuggest.js]: Using key !== "SORTED_SUGGESTIONS" will include other all other unwanted keys

File: unbxdAutosuggest.js (hunk 10 above)
Comment (divyadaglia-unbxd) [reply unbxdAutosuggest.js]: @ganeshparsads Done

File: unbxdAutosuggest.js (hunk 11 above)
Comment (divyadaglia-unbxd) [reply unbxdAutosuggest.js]: Done

File: unbxdAutosuggest.js (hunk 12 above)
Comment (divyadaglia-unbxd) [reply unbxdAutosuggest.js]: Done

File: unbxdAutosuggest.js (hunk 13 above)
Comment (divyadaglia-unbxd) [reply unbxdAutosuggest.js]: Created common customSort function

File: unbxdAutosuggest.js (hunk 9 above)
Comment (divyadaglia-unbxd) [reply unbxdAutosuggest.js]: Created common customSort function

File: unbxdAutosuggest.js (hunk 8 above)
Comment (divyadaglia-unbxd) [reply unbxdAutosuggest.js]: Created common customSort function

**Comment by divyadaglia-unbxd** [comment]
> @divyadaglia-unbxd could you please correct this PR.

**Pull request merged**